kernel/kalloc.c

```diff
@@ -72,11 +71,22 @@ kalloc(void)
 
   acquire(&kmem.lock);
   r = kmem.freelist;
-  if(r)
+  if (r)
     kmem.freelist = r->next;
   release(&kmem.lock);
 
-  if(r)
-    memset((char*)r, 5, PGSIZE); // fill with junk
-  return (void*)r;
+  if (r)
+    memset((char *)r, 5, PGSIZE); // fill with junk
+  return (void *)r;
+}
+
+uint64 free_mem()
+{
+  struct run *r;
+  uint64 memcount = 0;
+  for (r = kmem.freelist; r != 0; r = r->next)
```

Comment:
this works, can you think of a solution that would retrieve the answer even faster than this one?

Reply:
made the kernel keep track of the free count